### PR TITLE
add multi param grouping button

### PR DESCRIPTION
**Pull request by mashehu**
As discussed in #422 

First version:
Jul-06-2020 17-30-23

To-do:

- [x] write help text
- [x] filter option in modal?
- [x] Shift+click / cmd+click to bulk select parameters in the table
- [x] Display some explanation text in the table if there are no top-level params left
- [x] add "Select  all" button
- [x] add tooltips

### CODE REVIEW

File: public_html/json_schema_build.php [hunk 1]
```diff
@@ -304,7 +304,40 @@
             </div>
         </div>
     </div>
-
+    <!-- Moving multiple params into group modal -->
+    <div class="modal fade" id="multi_select_modal" tabindex="-1" role="dialog" aria-hidden="true">
+        <div class="modal-dialog modal-dialog-scrollable modal-xl" role="document">
+            <div class="modal-content">
+                <div class="modal-header">
+                    <h4>Move parameters into this <span></span></h4>
```
Comment (ewels) [inline public_html/json_schema_build.php:312]:
```suggestion
                    <h4>Move parameters into <span></span></h4>
```

File: public_html/json_schema_build.php [hunk 2]
```diff
@@ -304,7 +304,40 @@
...
+                    <button type="button" class="close" data-dismiss="modal" aria-label="Close">
+                      <span aria-hidden="true">&times;</span>
+                    </button>
+                </div>
+                <div class="modal-body">
+                    <table class="table table-sm">
+                        <thead>
+                            <tr>
+                            <th scope="col">Move</th>
```
Comment (ewels) [inline public_html/json_schema_build.php:321]:
```suggestion
                            <th scope="col" style="width:1%"></th>
```

File: public_html/json_schema_build.php [hunk 3]
```diff
@@ -304,7 +304,40 @@
             </div>
         </div>
     </div>
-
+    <!-- Moving multiple params into group modal -->
+    <div class="modal fade" id="multi_select_modal" tabindex="-1" role="dialog" aria-hidden="true">
+        <div class="modal-dialog modal-dialog-scrollable modal-xl" role="document">
```
Comment (ewels) [inline public_html/json_schema_build.php:309]:
```suggestion
        <div class="modal-dialog modal-dialog-scrollable modal-lg" role="document">
```

File: public_html/json_schema_build.php [hunk 4]
```diff
@@ -304,7 +304,40 @@
...
+                        <thead>
+                            <tr>
+                            <th scope="col">Move</th>
+                            <th scope="col">Parameter</th>
+                            <th scope="col">Description</th>
+                            </tr>
+                        </thead>
+                    <tbody>
+                    </tbody>
+                </table>
```
Comment (ewels) [inline public_html/json_schema_build.php:328]:
```suggestion
                        <tbody>
                        </tbody>
                    </table>
```

File: public_html/json_schema_build.php [hunk 5]
```diff
@@ -304,7 +304,40 @@
...
+                </div>
+                <div class="modal-footer">
+                    <div class="col text-right pr-0">
+                    <button type="button" class="btn btn-secondary" data-dismiss="modal">Close</button>
+                    <button type="button" class="btn btn-primary disabled" data-dismiss="modal" id="move_params">
+                        Move parameters
```
Comment (ewels) [inline public_html/json_schema_build.php:334]:
```suggestion
                        Move into <span></span>
```

File: public_html/assets/js/nf-core-schema-builder.js [hunk 6]
```diff
@@ -926,7 +926,78 @@ $(function () {
...
+        var group_el = $('.schema_group[data-id="' + id + '"] .card-body');
+        var selected_params = $('#multi_select_modal').find('.move_param:checked')
+        for (let i = 0; i < selected_params.length; i++) {
+            var p_id = $(selected_params[i]).data('id')
+            var row_el = $('.schema_row[data-id="' + p_id + '"]');
+            group_el.append(row_el)
+        }
+        schema_order_change();
+    }
+    )
```
Comment (ewels) [inline public_html/assets/js/nf-core-schema-builder.js:1000]:
```suggestion
    });
```

File: public_html/assets/js/nf-core-schema-builder.js [hunk 7]
```diff
@@ -926,7 +926,78 @@ $(function () {
...
+            if (schema['properties'][k].hasOwnProperty('properties')) {
+                continue
+            }
+            
 
+            // Add to the preview
+            params += `
+                <tr>
+                    <td>
+                        <input type="checkbox" aria-label="Move this parameter" class="move_param" data-id=`+k+`>
```
Comment (ewels) [inline public_html/assets/js/nf-core-schema-builder.js:971]:
```suggestion
                        <input type="checkbox" aria-label="Move this parameter" class="move_param" data-id=`+k+` id="group-move-`+k+`">
```

File: public_html/assets/js/nf-core-schema-builder.js [hunk 8]
```diff
@@ -926,7 +926,78 @@ $(function () {
...
+                    </td>
+                    <td>
+                        `+k+`
```
Comment (ewels) [inline public_html/assets/js/nf-core-schema-builder.js:974]:
```suggestion
                        <label for="group-move-`+k+`" class="text-monospace">`+k+`</label>
```

File: public_html/assets/js/nf-core-schema-builder.js [hunk 9]
```diff
@@ -926,7 +926,78 @@ $(function () {
...
+                    </td>
+                    <td>
+                        `+ schema['properties'][k].description +`
```
Comment (ewels) [inline public_html/assets/js/nf-core-schema-builder.js:977]:
```suggestion
                        <label for="group-move-`+k+`" class="small">`+ schema['properties'][k].description +`</label>
```

File: public_html/json_schema_build.php [hunk 10]
```diff
@@ -304,7 +304,40 @@
...
+                            <th scope="col">Description</th>
+                            </tr>
+                        </thead>
+                    <tbody>
+                    </tbody>
+                </table>
+                </div>
+                <div class="modal-footer">
+                    <div class="col text-right pr-0">
+                    <button type="button" class="btn btn-secondary" data-dismiss="modal">Close</button>
```
Comment (ewels) [inline public_html/json_schema_build.php:332]:
```suggestion
                    <button type="button" class="btn btn-secondary" data-dismiss="modal">Cancel</button>
```

File: public_html/assets/js/nf-core-schema-builder.js [hunk 11]
```diff
@@ -1189,6 +1260,9 @@ function generate_group_row(id, param, child_params){
                 <div class="col-auto align-self-center schema_row_config border-left">
                     <i class="fas fa-cog"></i>
                 </div>
+                <div class="col-auto align-self-center schema_group_move_params" alt-text="Move paramter(s) into this group">
```
Comment (ewels) [inline public_html/assets/js/nf-core-schema-builder.js:1263]:
```suggestion
                <div class="col-auto align-self-center schema_group_move_params" title="Move parameter(s) into this group">
```

**Comment by mashehu** [comment]
Yes, I thought we had tooltips once 🤔 definitely should have them for all buttons without text.
I updated the to-dos with your suggestions.

**Comment by mashehu** [comment]
I might have overdone it with the tooltips for the icons...

I am currently failing to get the shift/cmd multi-select working following this https://stackoverflow.com/a/659571/1696643, but I will keep trying 🙂

**Comment by mashehu** [comment]
Of course it worked after a coffee break. 🙃 
Anyway, I don't think the filter option is really needed for now, so I will remove the draft and WIP labels.

File: public_html/json_schema_build.php [hunk 12]
```diff
@@ -304,7 +305,45 @@
...
+                        </thead>
+                        <tbody>
+                        </tbody>
+                    </table>
+                    
+                </div>
+                <div class="modal-footer">
+                    <div class="btn-group" role="group">
+                        <button type="button" class="btn btn-light" id="select_all_params">Select all</button>
+                        <button type="button" class="btn btn-light" id="deselect_all_params">Deselect all</button>
```
Comment (ewels) [inline public_html/json_schema_build.php:335]: I actually didn't notice these until I read the code 🙄 

```suggestion
                        <button type="button" class="btn btn-outline-secondary" id="select_all_params">Select all</button>
                        <button type="button" class="btn btn-outline-secondary" id="deselect_all_params">Deselect all</button>
```

File: public_html/assets/js/nf-core-schema-builder.js [hunk 13]
```diff
@@ -926,7 +927,104 @@ $(function () {
...
+    //
+    
+    $('#multi_select_modal').on('change', '.select_param', function(){
+        var num_selected = $('#multi_select_modal').find('.select_param:checked').length;
+        if(num_selected>0){
+            $('#multi_select_modal #move_params').removeClass("disabled");
+            if(num_selected === 1){
+                $('#multi_select_modal #move_params').html("Move 1 parameter");
+            } else {
+            $('#multi_select_modal #move_params').html("Move "+num_selected+" parameters");
```
Comment (ewels) [inline public_html/assets/js/nf-core-schema-builder.js:941]:
```suggestion
                $('#multi_select_modal #move_params').html("Move "+num_selected+" parameters");
```

File: public_html/json_schema_build.php [hunk 14]
```diff
@@ -304,7 +305,45 @@
...
+                    </div>
+                    <div class="col text-right pr-0">
+                    <button type="button" class="btn btn-secondary" data-dismiss="modal">Cancel</button>
+                    <button type="button" class="btn btn-primary disabled" id="move_params">
+                        Move into <span></span>
```
Comment (ewels) [inline public_html/json_schema_build.php:340]: My intention with this new `<span>` was to have the group name in the button, where you actually click. This makes the intention a bit more explicit.

However, thinking about it now the group names can be quite long so I disagree with my past self. So should probably revert this..

```suggestion
                        Move parameters
```

File: public_html/assets/js/nf-core-schema-builder.js [hunk 15]
```diff
@@ -926,7 +927,104 @@ $(function () {
...
+                <tr>
+                    <td>
+                        <input type="checkbox" aria-label="Move this parameter" class="select_param" data-id=`+k+` id="group-move-`+k+`">
+                    </td>
+                    <td>
+                        <label for="group-move-`+k+`" class="text-monospace">`+k+`</label>
+                    </td>
+                    <td>
+                        <label for="group-move-`+k+`" class="small">`+ schema['properties'][k].description +`</label>
+                    </td>
```
Comment (ewels) [inline public_html/assets/js/nf-core-schema-builder.js:977]: Minor, I just find this a little easier to read..

```suggestion
                    <td><input type="checkbox" aria-label="Move this parameter" class="select_param" data-id=`+k+` id="group-move-`+k+`"></td>
                    <td><label for="group-move-`+k+`" class="text-monospace">`+k+`</label></td>
                    <td><label for="group-move-`+k+`" class="small">`+ schema['properties'][k].description +`</label></td>
```

File: public_html/assets/js/nf-core-schema-builder.js [hunk 16]
```diff
@@ -926,7 +927,104 @@ $(function () {
...
+                </tr>
+                `
+        }
+        if (params===''){
+            params = '<div class="alert alert-info">No ungrouped parameters available.</div>';
+            $('#multi_select_modal .table').remove();
```
Comment (ewels) [inline public_html/assets/js/nf-core-schema-builder.js:983]: `remove()` breaks this if the user moves any params back to the top level. It shows the "no ungrouped" message forever more...

Better to have the alert in the main HTML but hidden by default, then just toggle their visibilities..

File: public_html/assets/js/nf-core-schema-builder.js [hunk 17]
```diff
@@ -1068,7 +1166,7 @@ function generate_param_row(id, param){
         <div class="col-auto align-self-center schema_row_grabber border-right">
             <i class="fas fa-grip-vertical"></i>
         </div>
-        <button class="col-auto align-self-center param_fa_icon ">`+fa_icon+`</button>
+        <button class="col-auto align-self-center param_fa_icon " title="Select icon" data-toggle="tooltip">`+fa_icon+`</button>
```
Comment (ewels) [inline public_html/assets/js/nf-core-schema-builder.js:1169]: We need to initialise these tooltips when new rows are added to the schema. I thought this was broken at first as I was starting with an empty schema and adding stuff.

Just needs a `row.find('[data-toggle="tooltip"]').tooltip()` somewhere.. (Let me know if you can't find it right away and I can take a look).

**Comment by ewels** [comment]
It upsets me slightly that the shift-clicking on the labels doesn't work for me (checkboxes are fine). After a very quick play I think that this might be down to a browser bug though rather than anything else though, so not worth messing with.

Could be worth adding a `class="small text-muted"` bit of help text above or below the table saying that you can shift-click the checkboxes though.. 👍

File: public_html/assets/js/nf-core-schema-builder.js (hunk 17 above)
Comment (mashehu) [reply public_html/assets/js/nf-core-schema-builder.js:1169]: Is it okay if I put that into the `init_group_sortable()` function? Would be called at the right time, but also a bit of a function overload...

Comment (ewels) [reply public_html/assets/js/nf-core-schema-builder.js:1207]: Hmm, I think that will not work for when new rows are added though (only groups). I would probably add it to the row / group generation code, ie. [here](https://github.com/nf-core/nf-co.re/blob/7c782c12fd6f8fa231970b30e5c5c3b9c2c84b48/public_html/assets/js/nf-core-schema-builder.js#L169) and [here](https://github.com/nf-core/nf-co.re/blob/7c782c12fd6f8fa231970b30e5c5c3b9c2c84b48/public_html/assets/js/nf-core-schema-builder.js#L193).

I guess it should just be something like this?

```javascript
param_row.prependTo('#schema-builder').find('.param_id').select();
param_row.tooltip();
```

File: public_html/assets/js/nf-core-schema-builder.js (hunk 16 above)
Comment (mashehu) [reply public_html/assets/js/nf-core-schema-builder.js]: fixed in 36e9458

File: public_html/assets/js/nf-core-schema-builder.js (hunk 17 above)
Comment (mashehu) [reply public_html/assets/js/nf-core-schema-builder.js:1192]: done 0ccb5b9

File: public_html/assets/js/nf-core-schema-builder.js [hunk 18]
```diff
@@ -926,6 +929,124 @@ $(function () {
...
+            if (schema['properties'][k].hasOwnProperty('properties')) {
+                continue
+            }
+            // create row for the table
+            params += `
+				<tr data-id=`+ k + `>
+                    <td><input type="checkbox" aria-label="Move this parameter" class="select_param" data-id=`+ k + ` id="group-move-` + k + `"></td>
+                    <td><label for="group-move-`+ k + `" class="text-monospace">` + k + `</label></td>
+                    <td><label for="group-move-`+ k + `" class="small">` + schema['properties'][k].description +`</label></td>
+                </tr>
```
Comment (ewels) [inline public_html/assets/js/nf-core-schema-builder.js:1002]: Missing the end of this escaped block, causing a JS error..

```suggestion
                </tr>`;
```

File: public_html/assets/js/nf-core-schema-builder.js [hunk 19]
```diff
@@ -926,6 +929,124 @@ $(function () {
...
+        $("#search_parameters").val("");
+        var params = '';
+        for (k in schema['properties']) {
+            // skip over Groups
+            if (schema['properties'][k].hasOwnProperty('properties')) {
+                continue
+            }
+            // create row for the table
+            params += `
+				<tr data-id=`+ k + `>
```
Comment (ewels) [inline public_html/assets/js/nf-core-schema-builder.js:998]:
```suggestion
            params += `<tr data-id=`+ k + `>
```

**Comment by ewels** [comment]
Made a few minor tweaks in b91e4c5e49cda19f1abf5728c93660e431da9503:

* My `param_row.tooltip();` guess didn't work. Changed how the tooltips are initialised for the whole website instead, now works as expected.
* Changed the table filter a little to use the `keyup` instead of `input propertychange`. Seems to work better for me now.
* Small general code cleanup (simplified one or two small bits, couple of new comments, moved a chunk of code to follow a more logical order)

Thanks for this, it's a great addition!

**Pull request merged**